staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -788,6 +788,59 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-preserve-unknown-fields"), *schema.XPreserveUnknownFields, "must be true or undefined"))
 	}
 
+	if schema.XListType != nil && schema.Type != "array" {
+		if len(schema.Type) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("type"), "must be array if x-kubernetes-list-type is set"))
+		} else {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("type"), schema.Type, "must be array if x-kubernetes-list-type is set"))
+		}
+	}
+
+	if schema.XListType == nil && schema.Type == "array" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("x-kubernetes-list-type"), "must be set if type is array"))
+	}
+
+	if schema.XListType != nil && *schema.XListType != "atomic" && *schema.XListType != "set" && *schema.XListType != "map" {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("x-kubernetes-list-type"), *schema.XListType, []string{"atomic", "set", "map"}))
+	}
+
+	if len(schema.XListMapKeys) > 0 && (schema.XListType == nil || *schema.XListType != "map") {
+		if schema.XListType == nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("x-kubernetes-list-type"), "must be map if x-kubernetes-list-map-keys is non-empty"))
+		} else {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-list-type"), *schema.XListType, "must be map if x-kubernetes-list-map-keys is non-empty"))
+		}
+	}
+
```

Comment:
up to you, but you might consider moving the next four checks into a block like this and factoring out the XListType == map check (I might be alone, but I'm going slightly cross-eyed with the strings of nil checks)

```go
if schema.XListType != nil && *schema.XListType == "map" {
  if len(schema.XListMapKeys) == 0 {
    ... required error
  }
  if schema.Items != nil && schema.Items.Schema == nil {
    ... invalid error
  }
  if schema.Items != nil && schema.Items.Schema != nil && schema.Items.Schema.Type != "object" {
    ... invalid error
  }
  if schema.Items != nil && schema.Items.Schema != nil && schema.Items.Schema.Type == "object" {
    ... check keys/properties
  }
}
```

Reply:
I also didn't see an error if `schema.Items == nil` for a map list... did I miss that?

Reply:
did you discuss to actually validate the `XListMapKeys` values as field names in the schema? Both that they exist and that they are scalars.